BUILD-406: shared resource validating admission EP

gabemontero opened the pull request:
EP that stems from early testing and demoing of the use of RHEL entitlements in OpenShift Builds via the new Shared Resources CSI Driver and Controller, which is tech preview in 4.10

This EP proscribes 3 major validations around the creation/modification of SharedConfigMaps and SharedSecrets, as well as the use of CSI volumes provisioned by the Shared Resource CSI Driver, via new k8s ValidatingAdmissionWebhooks.

While similar validations currently or will exist in the controller / CSI driver, use of the webhook provides better usability when dealing with violations of the validations.

jsafrane commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,407 @@
+ 1. Is there a "Pod Spec'able" filter wrt configuring validating admission webhooks, so we catch all the subtypes of Pod?  Or do we have to list each k8s type that has Volumes in the spec?  Do not think there is a "pod spec'able" filter, but let's confirm.
+ 2. Existing k8s metrics cited below are sufficient for execution times for each request/response interaction when our new admission validations
+    are called, correct? If not, does it make sense to have new metrics be histograms around the admission validations execution times?
+ 3. Validating webhooks can define / maintain metrics and a metric service, right?
```
Yes, from metrics perspective, a webhook it's just a Pod that exposes a metric endpoint.

gabemontero replied:
thanks for confirmation

jsafrane commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,407 @@
+But if the API Server is not up to honor those requests, then any requests from the kubelet to mount our volumes will be denied.
+But that is no different than the level of service provided in 4.10.
+
+### Version Skew Strategy
```
Until the CSI driver is behind I don't think it's necessary to add convoluted conditions to validate existing Shared Resources, there can't be any version skew.

gabemontero replied:
Per my agreement at https://github.com/openshift/enhancements/pull/1056#discussion_r823895693 I will clean this up too

jsafrane commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,407 @@
+ 2. Existing k8s metrics cited below are sufficient for execution times for each request/response interaction when our new admission validations
+    are called, correct? If not, does it make sense to have new metrics be histograms around the admission validations execution times?
+ 3. Validating webhooks can define / maintain metrics and a metric service, right?
+ 4. Since we are going to have backup validations continue in the controller, can the webhook and controller share metrics, or do we need them to maintain separately named, but logically equivalent metrics?
```
They can share metrics, still, I am not sure there is a need for the controller + extra conditions.

gabemontero replied:
Yeah aside from not worrying about 4.10 customer instances with "openshift-*" names, the EP template does asks questions around "what happens if the new component is down" and I do like the idea of the 4.11 version of the controller being able to enforce our restrictions if the webhook is inexplicably down for some reason, but enough of the API server is available so that pods with volume mounts to our driver can still be started.

Let's revisit, as well as see what the other reviewers say, once I make the simplifications around the "openshift-*" naming scenario.

Thanks.

jsafrane commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,407 @@
+
+### Open Questions [optional]
+
+ 1. Is there a "Pod Spec'able" filter wrt configuring validating admission webhooks, so we catch all the subtypes of Pod?  Or do we have to list each k8s type that has Volumes in the spec?  Do not think there is a "pod spec'able" filter, but let's confirm.
```
I wish I knew the answer. IMO, PodSpec can be embedded in CRDs / external API servers and there is no easy way to find them.

gabemontero replied:
yeah we'll see if any of the other assignees to this PR know ... thanks

adambkaplan replied:
Unfortunately I don't think there is a notion of "duck typing" APIs for webhooks - you need to specify the top level APIs to listen to. That would mean we watch things that we ship in OpenShift that reference pod volumes

- Pod itself
- Basic workload types (Deployment, StatefulSet, Job, CronJob, DaemonSet)
- More "inner" workload types (ReplicaSet, ReplicationController)?
- BuildConfig and Build

gabemontero replied:
thanks for the additional confirmation @adambkaplan as well ascalling out duck typing, as that is what in part prompted my thinking ... I should have used that phrase as well (just forgot).

Maybe I'll add that phrase at some point, but more than likely I just remove this questions and state the reality.  That said, I should also replace this question with bubbling up the admission webhook vs. plugin question, per your other comments wrt the concerns you and I both voiced on the impact of a misbehaving webhook.

thanks again

gabemontero commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,407 @@
+
+The "openshift-*" naming restriction is new as compared to the initial tech preview release of Shared Resources in 4.10.
+
+A ValidatingAdmissionWebhook will be used to inspect all created Pods, and k8s Pod derived types (Deployments, DaemonSets, etc.), 
+as well Builds and BuildConfigs, and make sure no read-write Shared Resource CSI Volumes are specified.
```
yep although I cannot find any linked references, there have been discussions around this idea

*IF* I recall correctly, I thought we received a general warning from @deads2k that we should stay away from mutating webhooks

Now, if I'm not remembering the outcome of the conversation correctly, or if @deads2k 's stance has changed, or if there are exceptions to that stance which this scenario falls under, then yes, just mutating a volume specs to insure `readOnly: true` for use of the Shared Resource CSI driver does have the feel of "helping usability" even further.

@deads2k - WDYT?

thanks

gabemontero commented:
ok @jsafrane I have pushed a new commit that is mostly my attempt to address your comments that I immediately signed off on (the mutating one is still pending for me at least on some feedback from @deads2k based on what I think he has told me in the past)

There are also some minor tweaks based on my taking another pass at reinterpreting the EP template.

thanks

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+- REMINDER: the status structs of Shared Resources have an array of Conditions.  See 
+[here](https://github.com/openshift/api/blob/dbc82f2a3bc8b07e39628d36aee19d5a62847e2c/sharedresource/v1alpha1/types_shared_configmap.go#L83-L88) and
+[here](https://github.com/openshift/api/blob/dbc82f2a3bc8b07e39628d36aee19d5a62847e2c/sharedresource/v1alpha1/types_shared_secret.go#L81-L86)
+- The components who create pre-populated Shared Resources should have a controller that performs whatever validations
+they deem possible to verify the References, and those controllers are what will update the Status Condition arrays
+for the pre-populated Shared Resources
```
It seems odd to me that a separate controller would need to update status. What status conditions will these controllers set?

gabemontero replied:
Yeah I need to do a better job of describing what @deads2k and I discussed.  I'll post a concrete example here (maybe I even put some form of that in the EP).  Assuming that concrete example helps clear things up for you @adambkaplan, I'll then move on to improving the EP itself.

So consider:
- the Insights Operator starts creating a SharedSecret to encapsulate the entitlement secret they maintain
- they seem to already being doing some "validations" around that secret 
- maybe then even have more in the works?
- maybe they agree to do some items that we or @deads2k suggest?
- if they discover something wrong with the entitlement secret, the insights operator updates the status of the SharedSecret they create for using it, setting true/false, reason, message on our well known "validity" condition, explaining someting is wrong
- Very similar to what CVO operators do with the well known conditions CVO defines for the ClusterOperator object
- then OpenShift Build development, we we get the case that says "entitled builds don't work", will ask them to check the conditions on the SharedSecret that Insights Operator creates
- the Reason/Message there helps the customer figure out what he needs to do to fix his entitlement credentials (maybe he has to log onto console.redhat.com and do something) and then his entitlement builds work

Does that help @adambkaplan ?  If it does, including perhaps you accepting the CVO operator analogy, do you think you are on board?

thanks

adambkaplan replied:
Now that I see the full impact of my idea above, I'm less convinced adding something to `spec` is a good idea. Leaving `Ready` as an unmanaged condition that the CSI driver checks is perhaps the best option. We would want to document this in the status API.

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+corresponding ConfigMap with the necessary information.  This is akin to how we create in-payload image streams via the PRs against the 
+samples operator today.
+
+There will be a "ocp-sharedsecret-list" and "ocp-sharedconfigmap-list" ConfigMap.  The key of an entry is the Shared Resource
+name.  The value should be of the form "namespace:name" in order to capture. 
```
May help to provide an example here with the Insights operator's cluster entitlement:

```
kind: ConfigMap
apiVersion: v1
metadata:
  name: ocp-sharedsecret-list
  namespace: <namespace of the webhook>
data:
  openshift-etc-pki-entitlement: openshift-config-managed:etc-pki-entitlement
```

gabemontero replied:
will do in next push

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+Secret or ConfigMap.
+
+A [ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook)
+will be used to inspect all created/modified Shared Resources, and make sure their names do not start
```
Inspecting object updates could be problematic. Let's say an item is _removed_ from our allowlist on upgrade. Then we could have a `SharedSecret` or `SharedConfigMap` that is valid on creation, but is invalid on update after the upgrade removes the .

I don't think we want to block object updates if it already exists.

gabemontero replied:
good point .... I'll reduce to create and take out the modify bit

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+corresponding ConfigMap with the necessary information.  This is akin to how we create in-payload image streams via the PRs against the 
+samples operator today.
+
+There will be a "ocp-sharedsecret-list" and "ocp-sharedconfigmap-list" ConfigMap.  The key of an entry is the Shared Resource
```
We should also note that these lists will be managed by the CVO. Users will not be able to change them without the CVO restoring the original allowlists.

gabemontero replied:
I am pretty sure I do that elsewhere, but will cross reference with this line, what the EP template says, and reiterate here in some form of fashion to make sure it is clear.

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+### Implementation Details/Notes/Constraints [optional]
+
+Generally, basic field checking of the API Object created or modified when the webhook is called should be sufficient
+for the 'readOnly: true' and 'openshif-*' restriction enforcement.
```
nit:
```suggestion
for the 'readOnly: true' and 'openshift-*' restriction enforcement.
```

gabemontero replied:
cool thanks

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+= SharedSecret creation/modification failures because of an openshift-* name.  Metric labels for share name, secret name, secret namespace
+- SharedConfigMap creation/modification failures because of an openshift-* name.  Metric labels for share name, configmap name, configmap namespace
+- SharedConfigMap/SharedSecret having the validity condition go to invalid.  Metric labels for share name, reference name and namespace
+- Pod specable creation/modification failure for Shared Resource CSI Volumes without 'readOnly: true'.  Metric labels for namespace, Group/Kind/Version, share name.
```
We absolutely need a response time histogram for the webhook if we are reacting all pods, possibly with separate labels for accepted/rejected. If SD team gets a report that it is taking a long time for pods to be created, we need to see to rule out that this webhook was the cause. SLO would need to be something like "99% of all requests return a response within 1 second."

gabemontero replied:
ok cool thanks for the confirmation

when we do get down to implementation, we should work with metrics team wrt a new metric vs. the generic webhook one I found

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+Deferring on any metrics around associated Secret/ConfigMap validation until some of the open questions around that topic
+are answered.
+
+Alert 1:  using the `rate()` prometheus function, we can compare our failure counts above with existing metrics from 4.10
+wrt successful mounts, and build alerts where, for example, if failures / successes over 1 minute > 0.1, fire.  A pseudo-code
+example: `rate(read_only_failures[1m] / rate(successful_mount[1m]) > 0.1`.  Debating info only to moderate priority for the alert
```
Is there a general metric for volume mounts succeeding/failing?

gabemontero replied:
I have not found one.   I'll look again.  If still no luck, I'll add an open question, reach out to the storage team.

gabemontero replied:
I will be posting what I did find as part of an open question on this topic.... summary, "sort of" but it is only alpha level at this time https://kubernetes-csi.github.io/docs/volume-health-monitor.html

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+with the CLI invocation used, and any associated YAML for the object, should lead to straight forward diagnostics for 
+support.
+
+The enforcement of 'readOnly: true' in particular is worth discussing in this section.  It seems like very simple 
+pattern matching / field checking, but unanticipated bugs when performing this check
+in a ValidatingAdmissionWebhook could be impactful, as we are ultimately gating whether a Pod is allowed or not.  If 
+there is a bug which results in valid Pods being rejected, that could have broad and major consequences.
```
If we are enforcing things at the `Pod` level, then ultimately every workload will run through this check.
This could potentially lead to undesirable behavior if the workload managing the pod does not gracefully handle repeated failed pod creation.

The same issue will apply to workload APIs that the webhook is not aware of. For instance, a Tekton TaskRun may create a faulty Pod and not report this error in a friendly manner.

gabemontero replied:
OK so I take it this comment is plus one'ing my concern @adambkaplan :-)

Your point about admission plugins vs. webhooks and getting back to the choice item that came up with the KEP are key.  Presumably with the reviewers here between you, me, @deads2k and @jsafrane will reach a final consensus.

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+    will show >1s latency and alert `WebhookAdmissionLatencyHigh` will fire.
+- Otherwise, we should be able to fall back on must gather.
+
+##### Disable the API extension (e.g. remove ValidatingWebhookConfiguration `<names of the our new webhook configs>`)
```
Is this achievable - won't our operator deploy the webhook and reconcile the `ValidatingWebhookConfiguration` object?

gabemontero replied:
Well I'm actually just following the template, trying to address the questions it asks.  I just to converted a bullet point in the template to a heading for readability.

Can't seem to do line numbers when referencing lines in markdown files, but if you search for `disable the API extension (e.g. remove MutatingWebhookConfiguration xyz, remove APIService foo)` in https://github.com/openshift/enhancements/blob/db2ce78a020c2998096f7763bddbe108e40ce3ca/guidelines/enhancement_template.md you'll see the question.

As to "achievability", other than disabling the CSO (I imagine that has the same consequence support wise as disabling the CVO, but don't know that with certainty) is the only means for that I can think of.

Maybe for crazy critical support situations we will suggest that to customers to prevent clusters from going belly up ??

adambkaplan replied:
This might be the best option we have in terms of "emergency support":
- Scale the CVO deployment to 0
- Scale the cluster-storage-operator deployment to 0
- Remove the ValidatingAdmissionWebhook configuration.

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+
+Initial submission of this EP
+
+## Drawbacks
```
Adding a webhook that watches all pod creation risks cluster stability.

gabemontero replied:
I mentioned that concern earlier but sure I can mention it here, though per the other thread, maybe we switch on that for the Pod validations.

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+
+So we do not envision needing new repositories or a new operator for managing the webhook(s).
+
+The current core repository of [https://github.com/openshift/csi-driver-shared-resource](https://github.com/openshift/csi-driver-shared-resource)
```
+1 to this

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,413 @@
+
+## Alternatives
+
+First, currently see no reasons to use the older method (like in the early 3.0 days) of an admission plugin inside openshift-apiserver
+for validation, like you see for older API like Builds, DeploymentConfigs, ImageStreams, and Templates, could be found.
+But of course any SME guidance around this is welcome.
```
I think a reason to use an admission plugin instead of webhook is the fact that we will be validating pods. Validating pods requires a level of availability that we already guarantee with openshift-apiserver. Openshift-apiserver already has a mutating admission webhook for imagestreams, this would not be that much different.

When I was working on the draft KEP for CSI volume security, we went down the path of adding an admission plugin to k8s precisely because we were validating pods. We can use feature gates to enable/disable the admission plugin.

gabemontero replied:
Right (forgot that precise element of the plugin/webhook comparison).  So a couple of follow ups @adambkaplan :
1) does it make sense to use validating webhooks for the SharedResource openshift-* name validation, or just use an admission plugin for both
2) I know we have examples of such admission plugins for the OpenShift API types in the OpenShift API server today, but I don't remember in the KEP CSI volume discussions, not that I was involved in all of them, if there is an example of OpenShift injecting one for a K8 type ... i.e. injecting one in the k8s API server.  Is that maybe how we prevent namespace creation from the command line with names starting with "openshift-" ?  Do you recall @adambkaplan 

I did not bring up this question with @deads2k during my exchange with him a few days ago.

gabemontero commented:
@adambkaplan - first pass at updates based on your comments from last week
@jsafrane - I also added the "should we be mutating" item to the open questions

unless @deads2k manages to find time to look at this in the next few days, I'll try to reach out to him in slack on some of the open questions that are either a) I need a quick refresh on conversations a few months ago, or b) he is most like the most authoritative opinion

gabemontero commented:
although github per their status page is a bit compromised as I type, FYI I just pushed a new commit with metrics related updates per slack conversations I've had with the monitoring team and some of the reviewers here.

I don't see the new commit registered yet, but presumably it will show up eventually.  I'll monitor throughout the day and repush if need be.

gabemontero commented:
@adambkaplan @jsafrane @deads2k I've updated the EP based on the exchanged @deads2k and I had in https://coreos.slack.com/archives/CE4L0F143/p1647612099320099

With that exchanged, I've settled on the validating webhook implementation choice, along with adding needed details on the associated configuration I picked up form @deads2k.  As part of this, I pruned a good chunk of the open questions, and moved some of the comments here in the PR along those open questions to the alternatives section.

PTAL - thanks

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,486 @@
+#### New SLIs for the new admission validations to help cluster administrators and support
+
+Separate metrics for
+= SharedSecret creation/modification failures because of an openshift-* name.  Metric labels for share name, secret name, secret namespace
+- SharedConfigMap creation/modification failures because of an openshift-* name.  Metric labels for share name, configmap name, configmap namespace
+- SharedConfigMap/SharedSecret having the validity condition go to invalid.  Metric labels for share name, reference name and namespace
+- Pod specable creation/modification success/failure count for Shared Resource CSI Volumes without 'readOnly: true'.  Metric labels for namespace, Group/Kind/Version, share name, 
```
These metrics all have unbound cardinality. Is this a concern?

gabemontero replied:
sure, I took the fork in the road around "optimal debuggability" from a support perspective, but do not feel super strong about it either.

But I'm fine with removing the labels for each of them, and making them global in nature.  Still has value in that alerts based of them growing will trigger the need for more debug.

I'll proceed removing those labels @adambkaplan but certainly chime in with any "meet in the middle labeling" you like.

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,486 @@
+#### Measuring / Verifying impact on existing API Server SLIs
+
+From existing metric discovery, what we will depend on:
+- the EP template example in "Support Procedures" of using `webhook_admission_duration_seconds("<name of our new admission controller>", "validating", "put", "false")`, with >1s latency resulting in the `WebhookAdmissionLatencyHigh` alert firing, appears to be the valid choice addressing our new webhooks' impact on the API Object create/modify flows.  Is that correct?
```
This feels correct - if our admission controller has high latency, we fire an alert.

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,486 @@
+...
+webhooks:
+  - name: podsharedresourcereadonlywebhook.openshift.com
+    objectSelector:
```
Be sure to add the `failurePolicy` - any timeout/500 errors should be ignored ("fail open")
```suggestion
    failurePolicy: Ignore # Do not fail object creation if this webhook times out or returns 5xx errors
    objectSelector:
```

See https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#validatingwebhook-v1-admissionregistration-k8s-io

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,486 @@
+
+And provide validation failures on API object creation/modification for each of these whenever possible.
+
+### Non-Goals
```
Perhaps add non-goals around the "is the underlying object valid" feature.

gabemontero commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,486 @@
+        - {key: openshift.io/run-level, operator: NotIn, values: ["0", "1"]} 
+        - {key: openshift.io/exempt-from-shared-resource-readonlytrue-webhook, operator: NotIn, values: ["true"]}
+    rules:
+      - operations: [ "CREATE" ]
+        apiGroups: []
+        apiVersions: []
+        resources: [ "pods", "deployments", "daemonsets", ....]
+        scope: "*"
```
good catch @adambkaplan thanks

my next push will include:
- I've split out apps from pods in my pseudo yaml here
- I also added openshift deploymentconfig as a separate item; yes, they are not "preferred", but they still exist and are used
- I added an open question for your query to @soltysh in case he does not get to it before I push the update
-  I added him as a reviewer in the doc
- I'll assign him to the PR when I've pushed the changes

gabemontero commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,486 @@
+- The first `matchExpression` is the means of preventing the use of our new webhook for any Pods of the core OCP control pane.  Because of this
+label selector we will filter out components like [the K8s API Server Operator Namespace manifest](https://github.com/openshift/cluster-kube-apiserver-operator/blob/d80b406ec4844328e04253f9ed5aec3f4a29834a/manifests/0000_20_kube-apiserver-operator_00_namespace.yaml#L11)
+- The second `matchExpression` is the means of preventing the use of our new webhook for any Pods in other OCP namespaces we want to exclude.
+Initially, we are thinking only the [CSI Drivers namespace](https://github.com/openshift/cluster-storage-operator/blob/2a3fbb68afd066d16582b09f9fd834c7fd03457e/manifests/02_csi_driver_operators_namespace.yaml) 
+will be excluded (as that is where we will also host our webhook), so
```
it is more the  latter, but to be clear, with the use if `NotIn`, we would be validating by default, except for pods for namespaces in the `NotIn` list .... for the detailed walk through, revisit the thread between David and me at  https://coreos.slack.com/archives/CE4L0F143/p1647612099320099 and in particular the existing example he pointed me to https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/manifests/0000_20_kube-apiserver-operator_00_namespace.yaml#L11 and his comments https://coreos.slack.com/archives/CE4L0F143/p1647615292929869?thread_ts=1647612099.320099&cid=CE4L0F143

adambkaplan replied:
I found it - the admission controller config has a `namespaceSelector` option. This is how we ignore pods in these namespaces: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-namespaceselector

gabemontero commented:
updates for you comments pushed @adambkaplan 

/assign @soltysh

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,532 @@
+...
+webhooks:
+  - name: pod.csi.sharedresource.openshift.com
+    failurePolicy: Ignore
```
Also add `namespaceSelector:`:

```suggestion
    failurePolicy: Ignore
    namespaceSelector:
      matchExpressions:
        - key: runlevel
          operator: NotIn
          values: ["0", "1"]
        - key: openshift.io/run-level
          operator: NotIn
          values: ["0", "1"]
```

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,532 @@
+  - name: pod.csi.sharedresource.openshift.com
+    failurePolicy: Ignore
+    objectSelector:
+      matchExpressions:
+        - {key: openshift.io/run-level, operator: NotIn, values: ["0", "1"]} 
+        - {key: csi.sharedresource.openshift.io/skip-validation, operator: NotIn, values: ["true"]}
```
Drop the run-level match, covered by the namespace selector.

```suggestion
      matchExpressions:
        - {key: csi.sharedresource.openshift.io/skip-validation, operator: NotIn, values: ["true"]}
```

gabemontero replied:
actually I'm moving the skip-validation item to the namespace selector as well, as that was my intent:  allow for marking of  entire namespaces to be skipped, independent of the run level stuff.

adambkaplan commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,532 @@
+kind: ValidatingWebhookConfiguration
+...
+webhooks:
+  - name: pod.csi.sharedresource.openshift.com
```
nit - we use .io TLD within OpenShift
```suggestion
  - name: pod.csi.sharedresource.openshift.io
```

gabemontero commented:
updates from you comments yesterday pushed @adambkaplan

adambkaplan reviewed (approved):
/approve

Hitting approve on this EP. At this point we have vetted the configuration of the webhook and the scope of changes to be made. I'm ready to lgtm once the commits have been squashed.

gabemontero commented on enhancements/cluster-scope-secret-volumes/shared-resource-validation.md:
```diff
@@ -0,0 +1,524 @@
+example: `rate(read_only_failures[1m] / rate(successful_mount[1m]) > 0.1`.  Debating info only to moderate priority for the alert.
+We'll finalize during implementation and update EP accordingly.
+
+Alert 2: if any Shared Resources have an invalid Reference.  Info level priority for user created Shared Resources.  Considering
```
we can debate during implementation

gabemontero commented:
last suggestion pulled in and commits squashed @adambkaplan

adambkaplan commented:
/cc @coreydaley

adambkaplan reviewed (approved):
/lgtm
